18F/identity-idp
# LG-3420: Poll for status of document verification

**Why**: As a user, I want the client to tell the IDP to process images and poll the IDP for results of process images, so the upload process can be done async

**Implementation Notes:**

This pull request...

- Generates an encryption key to be reused in context and included in the final submission payload
- Encrypts values at `withBackgroundEncryptedUpload` with per-field IV
- Interprets a `202` response status code from the server as a "pending result"
- Schedules a `getStatus` request every 2.5s while result is pending

**Testing Procedure:**

Aside from automated tests, it's a bit difficult to test manually, due to the fact that presigned URLs are not yet implemented for local development, nor does the status check ever resolve to completed.

You can test parts of it by...

- Changing `doc_auth_enable_presigned_s3_urls` to `'true'` in `application.yml`
- Setting [S3 upload URLs](https://github.com/18F/identity-idp/blob/21be6a360593180a608324fb4b43306bf14ef9ed/app/views/idv/shared/_document_capture.html.erb#L13-L15) to dummy URLs in local copy of the branch

At which point, you can see...

- Selecting a file will issue an upload request to the given URLs, containing encrypted copy of the image
- Submitting the form sends a request to the `document_verification` step action URL, responding with status code 202 and with a payload containing only the following keys:
  - `front_image_iv`
  - `front_image_url`
  - `back_image_iv`
  - `back_image_url`
  - `selfie_image_iv`
  - `selfie_image_url`
  - `document_capture_session_uuid`
  - `locale`
  - `encryption_key`
  - `step`
- Every 2.5 seconds, you should see a subsequent `document_verification_status` request, containing an empty payload, and responding with status code 202
